solutions/cfgym-102694C.mdx

```diff
@@ -0,0 +1,238 @@
+
+## Implementation
+<LanguageSection>
+
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+// CodeSnippet{Kattio}
+public class slothNaptime {
+	public static final int MAXN = (int) 3e5 + 5;
+	public static final int LOGN = 20;
+	public static ArrayList<Integer>[] adj = new ArrayList[MAXN];
+	// anc[i][j] is the 2^j-th parent of i.
+	public static int[][] anc = new int[MAXN][LOGN];
+	public static int[] depth = new int[MAXN];
+
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio();
+		int N = io.nextInt();
+		for (int i = 0; i < MAXN; i++) {
+			adj[i] = new ArrayList<>();
+		}
+		for (int i = 0; i < N-1; i++) {
+			int a = io.nextInt();
+			int b = io.nextInt();
+			adj[a].add(b);
+			adj[b].add(a);
+		}
+
+		dfs(1, 0);
+		int Q = io.nextInt();
+
+		for (int i = 1; i <= Q; i++) {
+			int start = io.nextInt();
+			int end = io.nextInt();
+			int energy = io.nextInt();
+			int lca = LCA(start, end);
+			int path1 = depth[start] - depth[lca];
+			int path2 = depth[end] - depth[lca];
+			if (energy <= path1) {
+				// moves up from start towards the lca.
+				System.out.println(jump(start, energy));
+			} else if (energy <= path1 + path2) {
+				// moves up from start to the lca, then goes down towards end.
+				System.out.println(jump(end, path2 - (energy - path1)));
+			} else {
+				// reaches the end.
+				System.out.println(end);
+			}
+		}
+	}
+
+	public static void dfs(int node, int par) {
+		anc[node][0] = par;
+		depth[node] = depth[anc[node][0]] + 1;
+
+		// initializes binary jumping for the current node.
+		for (int lvl = 1; lvl < LOGN; lvl++) {
+			anc[node][lvl] = anc[anc[node][lvl - 1]][lvl - 1];
+		}
+		for (int next : adj[node]) {
+			if (next != par) {
+				dfs(next, node);
+			}
+		}
+	}
+
+	// jump(i, j) returns jth ancestor of node i.
+	public static int jump(int node, int level) {
+		for (int i = 0; i < LOGN; i++) {
+			if ((level & (1 << i)) > 0) {
+				node = anc[node][i];
+			}
+		}
+		return (node > 0) ? node : 1;
+	}
+
+    // lca(i, j) returns least common ancestor of nodes i and j.
+	public static int LCA(int a, int b) {
+		if (depth[a] < depth[b]) {
+			int tempA = a;
+			a = b;
+			b = tempA;
+		}
+
+		a = jump(a, depth[a] - depth[b]);
+		if (a == b) {
+			return a;
+		}
+		for (int i = LOGN - 1; i >= 0; i--) {
+			int newA = anc[a][i];
+			int newB = anc[b][i];
+			if (newA != newB) {
+				a = newA;
+				b = newB;
+			}
+		}
+
+		return anc[a][0];
+	}
+	
+
+}
+
+
+```
+
+</JavaSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAXN = 3e5;
+const int LOGN = 20;
+int depth[MAXN + 1];
+// anc[i][j] is the 2^j-th parent of i.
+int anc[MAXN + 1][LOGN];
+vector<int> adj[MAXN + 1];
+
+void dfs(int node, int par) {
+	depth[node] = depth[anc[node][0] = par] + 1;
```

Comment:
even though i get what this assignment means you might want to take the nested assignment out

Reply:
all the other editorials use the same assignment, but sure